modules/core/04-channel/v2/types/packet.go

```diff
@@ -2,10 +2,62 @@ package types
 
 import (
 	"crypto/sha256"
+	"strings"
+
+	errorsmod "cosmossdk.io/errors"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	host "github.com/cosmos/ibc-go/v9/modules/core/24-host"
 )
 
+func (p Packet) ValidateBasic() error {
+	if len(p.Data) != 1 {
+		return errorsmod.Wrapf(ErrInvalidPacket, "packet data length must be 1, got %d", len(p.Data))
+	}
+
+	for _, pd := range p.Data {
+		if err := host.PortIdentifierValidator(pd.SourcePort); err != nil {
+			return errorsmod.Wrap(err, "invalid source port ID")
+		}
+		if err := host.PortIdentifierValidator(pd.DestinationPort); err != nil {
```

Comment:
I think it may be desirable to allow empty source or destination ports (one sided applications), but I guess it is simple enough to change this check in that reality